File: KotlinLibraryX/library/src/main/java/com/lee/library/adapter/core/ProxyAdapter.java
package com.lee.library.adapter.core;

import android.annotation.SuppressLint;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.IntDef;
import androidx.annotation.IntRange;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.util.ArrayList;
import java.util.List;

/**
 *
 * @author jv.lee
 * @date 2019/5/20.
 */
public class ProxyAdapter extends RecyclerView.Adapter {
    private List<View> mHeaderViews = new ArrayList<>();
    private List<View> mFooterViews = new ArrayList<>();

    private final RecyclerView.Adapter mAdapter;

    public ProxyAdapter(RecyclerView.Adapter adapter) {
        if (adapter == null) {
            throw new IllegalArgumentException();
        }
        mAdapter = adapter;
        //设置图片闪烁，给所有item view添加tag
        mAdapter.setHasStableIds(mAdapter.hasStableIds());
        setHasStableIds(adapter.hasStableIds());
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        RecyclerView.ViewHolder viewHolder;
        final int type = ViewTypeSpec.getType(viewType);
        final int value = ViewTypeSpec.getValue(viewType);
        if (type == ViewTypeSpec.HEADER) {
            viewHolder = new FixedViewHolder(mHeaderViews.get(value));
        } else if (type == ViewTypeSpec.FOOTER) {
            viewHolder = new FixedViewHolder(mFooterViews.get(value));
        } else {
            viewHolder = mAdapter.onCreateViewHolder(parent, viewType);
        }
        return viewHolder;
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        if (holder instanceof FixedViewHolder) {
            ((FixedViewHolder) holder).onBind();
        } else {
            int adjPosition = position - mHeaderViews.size();
            mAdapter.onBindViewHolder(holder, adjPosition);
        }
    }

    @Override
    public int getItemCount() {
        return mHeaderViews.size() + mFooterViews.size() + mAdapter.getItemCount();
    }

    public int getHeaderCount() {
        return mHeaderViews.size();
    }

    @Override
    public int getItemViewType(int position) {
        final int numHeaderView = mHeaderViews.size();

        if (position < numHeaderView) {
            return ViewTypeSpec.makeItemViewTypeSpec(position, ViewTypeSpec.HEADER);
        }

        final int adjPosition = position - numHeaderView;
        final int itemCount = mAdapter.getItemCount();
        if (adjPosition >= itemCount) {
            return ViewTypeSpec.makeItemViewTypeSpec(adjPosition - itemCount, ViewTypeSpec.FOOTER);
        }

        int itemViewType = mAdapter.getItemViewType(adjPosition);
        if (itemViewType < 0 || itemViewType > (1 << ViewTypeSpec.TYPE_SHIFT) - 1) {
            throw new IllegalArgumentException("Invalid item view type: RecyclerView.Adapter.getItemViewType return " + itemViewType);
        }
        return itemViewType;
    }

    @Override
    public void onDetachedFromRecyclerView(@NonNull RecyclerView recyclerView) {
        mAdapter.onDetachedFromRecyclerView(recyclerView);
    }

    @Override
    public void onAttachedToRecyclerView(@NonNull RecyclerView recyclerView) {
        mAdapter.onAttachedToRecyclerView(recyclerView);
    }

    @Override
    public void unregisterAdapterDataObserver(@NonNull RecyclerView.AdapterDataObserver observer) {
        mAdapter.unregisterAdapterDataObserver(observer);
    }

    @Override
    public void registerAdapterDataObserver(@NonNull RecyclerView.AdapterDataObserver observer) {
        mAdapter.registerAdapterDataObserver(observer);
    }

    @Override
    public void onViewDetachedFromWindow(@NonNull RecyclerView.ViewHolder holder) {
        if (holder instanceof FixedViewHolder) {
            return;
        }
        mAdapter.onViewDetachedFromWindow(holder);
    }

    @Override
    public void onViewAttachedToWindow(@NonNull RecyclerView.ViewHolder holder) {
        if (holder instanceof FixedViewHolder) {
            return;
        }
        mAdapter.onViewAttachedToWindow(holder);
    }

    @Override
    public boolean onFailedToRecycleView(@NonNull RecyclerView.ViewHolder holder) {
        if (holder instanceof FixedViewHolder) {
            return false;
        }
        return mAdapter.onFailedToRecycleView(holder);
    }

    @Override
    public void onViewRecycled(@NonNull RecyclerView.ViewHolder holder) {
        if (holder instanceof FixedViewHolder) {
            return;
        }
        mAdapter.onViewRecycled(holder);
    }

    @Override
    public long getItemId(int position) {
        int adjPosition = position - mHeaderViews.size();
        if (adjPosition >= 0 && adjPosition < mAdapter.getItemCount()) {
            return mAdapter.getItemId(adjPosition);
        }

        return RecyclerView.NO_ID;
    }

    private boolean isFixedViewType(int viewType) {
        final int type = ViewTypeSpec.getType(viewType);
        return type == ViewTypeSpec.HEADER || type == ViewTypeSpec.FOOTER;
    }

    public void addHeaderView(View view) {
        if (!mHeaderViews.contains(view)) {
            mHeaderViews.add(view);
        }
    }

    public void removeHeaderView(View view) {
        mHeaderViews.remove(view);
    }

    public void addFooterView(View view) {
        if (!mFooterViews.contains(view)) {
            mFooterViews.add(0,view);
        }
    }

    public void removeFooterView(View view) {
        mFooterViews.remove(view);
    }

    static class ViewTypeSpec {
        static final int TYPE_SHIFT = 30;
        static final int TYPE_MASK = 0x3 << TYPE_SHIFT;

        static final int UNSPECIFIED = 0 << TYPE_SHIFT;
        static final int HEADER = 1 << TYPE_SHIFT;
        static final int FOOTER = 2 << TYPE_SHIFT;

        @IntDef({UNSPECIFIED, HEADER, FOOTER})
        @Retention(RetentionPolicy.SOURCE)
        @interface ViewTypeSpecMode {
        }

        static int makeItemViewTypeSpec(@IntRange(from = 0, to = (1 << TYPE_SHIFT) - 1) int value,
                                        @ViewTypeSpecMode int type) {
            return (value & ~TYPE_MASK) | (type & TYPE_MASK);
        }

        @SuppressLint("WrongConstant")
        @ViewTypeSpecMode
        static int getType(int viewType) {
            return (viewType & TYPE_MASK);
        }

        static int getValue(int viewType) {
            return (viewType & ~TYPE_MASK);
        }
    }

    public static class FixedViewHolder extends RecyclerView.ViewHolder {

        FixedViewHolder(View itemView) {
            super(itemView);
            setIsRecyclable(false);
        }

        void onBind() {

        }
    }

    public RecyclerView.Adapter getChildAdapter() {
        return mAdapter;
    }
}

File: KotlinLibraryX/library/src/main/java/com/lee/library/adapter/core/ProxyAdapter.kt
package com.lee.library.adapter.core

import android.view.View
import android.view.ViewGroup
import androidx.annotation.IntDef
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView

/**
 * RecyclerView 代理适配器，可通过该适配器添加header、footer
 *
 * @author jv.lee
 * @date 2019/5/20
 */
class ProxyAdapter(
    private val recyclerView: RecyclerView,
    private val adapter: RecyclerView.Adapter<RecyclerView.ViewHolder>,
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    private val mHeaderViews: ArrayList<View> = arrayListOf()
    private val mFooterViews: ArrayList<View> = arrayListOf()

    init {
        // 给所有item view添加tag 防止图片闪烁
        adapter.setHasStableIds(adapter.hasStableIds())
        setHasStableIds(adapter.hasStableIds())
        changeSpanSize()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val type = getType(viewType)
        val value = getValue(viewType)
        return when (type) {
            ViewTypeSpec.HEADER -> {
                FixedViewHolder(mHeaderViews[value])
            }

            ViewTypeSpec.FOOTER -> {
                FixedViewHolder(mFooterViews[value])
            }

            else -> {
                adapter.onCreateViewHolder(parent, viewType)
            }
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (holder is FixedViewHolder) {
            holder.onBind()
        } else {
            val adjPosition = position - mHeaderViews.size
            adapter.onBindViewHolder(holder, adjPosition)
        }
    }

    override fun onDetachedFromRecyclerView(recyclerView: RecyclerView) {
        adapter.onDetachedFromRecyclerView(recyclerView)
    }

    override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
        adapter.onAttachedToRecyclerView(recyclerView)
    }

    override fun unregisterAdapterDataObserver(observer: RecyclerView.AdapterDataObserver) {
        adapter.unregisterAdapterDataObserver(observer)
    }

    override fun registerAdapterDataObserver(observer: RecyclerView.AdapterDataObserver) {
        adapter.registerAdapterDataObserver(observer)
    }

    override fun onViewDetachedFromWindow(holder: RecyclerView.ViewHolder) {
        if (holder is FixedViewHolder) {
            return
        }
        adapter.onViewDetachedFromWindow(holder)
    }

    override fun onViewAttachedToWindow(holder: RecyclerView.ViewHolder) {
        if (holder is FixedViewHolder) {
            return
        }
        adapter.onViewAttachedToWindow(holder)
    }

    override fun onFailedToRecycleView(holder: RecyclerView.ViewHolder): Boolean {
        if (holder is FixedViewHolder) {
            return false
        }
        return adapter.onFailedToRecycleView(holder)
    }

    override fun onViewRecycled(holder: RecyclerView.ViewHolder) {
        if (holder is FixedViewHolder) {
            return
        }
        adapter.onViewRecycled(holder)
    }

    override fun getItemId(position: Int): Long {
        val adjPosition = position - mHeaderViews.size
        if (adjPosition >= 0 && adjPosition < adapter.itemCount) {
            return adapter.getItemId(adjPosition)
        }
        return RecyclerView.NO_ID
    }

    override fun getItemCount() = mHeaderViews.size + mFooterViews.size + adapter.itemCount

    override fun getItemViewType(position: Int): Int {
        val numHeaderView = mHeaderViews.size

        if (position < numHeaderView) {
            return makeItemViewTypeSpec(position, ViewTypeSpec.HEADER)
        }

        val adjPosition = position - numHeaderView
        val itemCount = adapter.itemCount
        if (adjPosition >= itemCount) {
            return makeItemViewTypeSpec(adjPosition - itemCount, ViewTypeSpec.FOOTER)
        }

        val itemViewType = adapter.getItemViewType(adjPosition)
        if (itemViewType < 0 || itemViewType > (1 shl TYPE_SHIFT) - 1) {
            throw IllegalArgumentException(
                "Invalid item view type: RecyclerView.Adapter.getItemViewType return $itemViewType"
            )
        }
        return itemViewType
    }

    fun addHeaderView(view: View) {
        if (!mHeaderViews.contains(view)) {
            mHeaderViews.add(view)
        }
    }

    fun removeHeaderView(view: View) {
        mHeaderViews.remove(view)
    }

    fun addFooterView(view: View) {
        if (!mFooterViews.contains(view)) {
            mFooterViews.add(view)
        }
    }

    fun addFooterViewAtTop(view: View) {
        if (!mFooterViews.contains(view)) {
            mFooterViews.add(0, view)
        }
    }

    fun removeFooterView(view: View) {
        mFooterViews.remove(view)
    }

    fun getHeaderCount() = mHeaderViews.size

    fun getFooterCount() = mFooterViews.size

    private fun isFixedViewType(viewType: Int): Boolean {
        val type = getType(viewType)
        return type == ViewTypeSpec.HEADER || type == ViewTypeSpec.FOOTER
    }

    private fun changeSpanSize() {
        val layoutManager = recyclerView.layoutManager
        if (layoutManager is GridLayoutManager) {
            layoutManager.spanSizeLookup = object : ProxySpanSizeLookup(recyclerView) {
                override fun buildSpanSize(position: Int): Int = 1
            }
        }
    }

    private class FixedViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        init {
            setIsRecyclable(false)
        }

        fun onBind() {}
    }

    private companion object {
        const val TYPE_SHIFT = 30
        const val TYPE_MASK = 0x3 shl TYPE_SHIFT

        fun makeItemViewTypeSpec(
            @androidx.annotation.IntRange(
                from = 0,
                to = ((1 shl TYPE_SHIFT) - 1).toLong()
            ) value: Int, @ViewTypeSpec type: Int
        ): Int {
            return (value and TYPE_MASK.inv()) or (type and TYPE_MASK);
        }

        fun getType(viewType: Int): Int {
            return (viewType and TYPE_MASK)
        }

        fun getValue(viewType: Int): Int {
            return (viewType and TYPE_MASK.inv())
        }
    }
}

@IntDef(ViewTypeSpec.UNSPECIFIED, ViewTypeSpec.HEADER, ViewTypeSpec.FOOTER)
@Retention(AnnotationRetention.SOURCE)
annotation class ViewTypeSpec {
    companion object {
        const val UNSPECIFIED = 0
        const val HEADER = 1 shl 30
        const val FOOTER = 2 shl 30
    }
}